_posts/2018-12-05-exploit-model.md

```diff
@@ -0,0 +1,153 @@
+
+```
++--------+              +---------+
+|  p1a   |              |   p2a   |                      APP
++--------+              +---------+
+<<< 300 syscalls  >>>   <<< 300 syscalls  >>>
++-------------------+   +-------------------+
+|      gvisor       |   |      gvisor       |            GVISOR
+|   (code exec)     |   |                   |            
++-------------------+   +-------------------+
+<<<  83 syscalls  >>>   <<<  83 syscalls  >>>     <--- Attack surface 
++--------------------------------------------------+
+|                    kernel                        |     KERNEL
++--------------------------------------------------+
+```
+
+Based on this, in order to break isolation between gvisor instances, an attacker would need to obtain code exec or something similar in the kernel via the 83 syscalls allowed.
+
+### Another model: LibOS with Nabla
+
+Let's compare the Nabla containers model with that of gvisor. The components are very similar, gvisor is a userspace unikernel, and Nabla has a Library OS (LibOS). This LibOS does a similar function - implementing the functionality of the kernel in userspace and only calling the kernel when necessary. 
+
+We note that right now, we don't distinguish between the application and the LibOS, and we treat the LibOS as untrusted. For the application and LibOS, we have a seccomp policy that only allows 7 syscalls, with only two file descriptions (a block device and tap device).
+
+```
++--------+              +---------+
+|  p1a   |              |   p2a   |                      APP
++-------------------+   +-------------------+
+|    nabla LibOS    |   |   nabla LibOS     |            LibOS
+|    (code exec)    |   |                   |            
++-------------------+   +-------------------+
+<<<   7 syscalls  >>>   <<<   7 syscalls  >>>     <--- Attack surface 
++--------------------------------------------------+
+|                    kernel                        |     KERNEL
++--------------------------------------------------+
+```
+
+For nabla, the attack surface to obtain code exec in the kernel would need to be from these 7 syscalls.
+
+
+## But it's a memory safe langauge!
+
+So gvisor does two things, it reduces the attack surface to the kernel, and has an protection boundary which gaurds against the inputs that can be invoked with the 83 syscalls (the gaurding is implicit in a way that the logic of the implementation would constrain the total number of syscalls invokable by a well behaved gvisor). Therefore, more work needs to be done by the attacker to bypass these gaurds.
+
+Some argue that gvisor is written in golang, and golang is memory safe. I believe that golang is a safer language from the PL standpoint, but it is hard to guarantee safety and should not be used as the only reason for security.
```

Comment:
I'd formulate in the direction of "golang is memory safe, assuming that golang runtime is correct".

Reply:
yea - that's true... (at least to the best of my knowledge)... 

I'm going to add an additional article to show another memory corruption via racing the interface implementation. But like you said, it's not according to the semantics for the language as per design.

I think something something Java memory safety too :) :) :)